_posts/2020-01-13-mixed-precision-training/mixed_precision.Rmd

```diff
@@ -0,0 +1,292 @@
+batch_size <- 32
+# 12895 is the number of items in the training set
+buffer_size <- 12895/batch_size
+
+# training dataset is resized, scaled to between 0 and 1,
+# cached, shuffled, and divided into batches
+train_dataset <- train %>%
+  dataset_map(function(record) {
+    record$image <- record$image %>%
+      tf$image$resize(size = c(256L, 256L)) %>%
+      tf$truediv(255)
+    record
+  }) %>%
+  dataset_cache() %>%
+  dataset_shuffle(buffer_size) %>%
+  dataset_batch(batch_size) %>%
+  dataset_map(unname)
+
+# test dataset is resized, scaled to between 0 and 1, and divided into batches
+test_dataset <- test %>% 
+  dataset_map(function(record) {
+    record$image <- record$image %>% 
+      tf$image$resize(size = c(256L, 256L)) %>%
+      tf$truediv(255)
+    record}) %>%
+  dataset_batch(batch_size) %>% 
+  dataset_map(unname)
+
+```
+
+In the above code, we cache the dataset after the resize and scale operations, as we want to minimize preprocessing time spent on the CPU.
+
+### Configuring MPT
+
+Our experiment uses Keras `fit` -- as opposed to a custom training loop --, and given these preconditions, running MPT is mostly a matter of adding three lines of code. (There is a small change to the model, as we'll see in a moment.)^[With custom training loops, losses should be scaled (multiplied by a large number) before being passed into the gradient calculation, to avoid numerical underflow/overflow. For detailed instructions, see the [documentation](https://tensorflow.rstudio.com/guide/keras/mixed_precision/).]
+
+We tell Keras to use the _mixed_float16_ `Policy`, and verify that the tensors have type `float16` while the `Variables` (weights) still are of type `float32`:
+
+```{r}
+# if you read this at a later time and get an error here,
+# check out whether the location in the codebase has changed
+mixed_precision <- tf$keras$mixed_precision$experimental
+
+policy <- mixed_precision$Policy('mixed_float16')
+mixed_precision$set_policy(policy)
+
+# float16
+policy$compute_dtype
+# float32
+policy$variable_dtype
+```
+
+The model is a straightforward convnet, with numbers of filters being multiples of 8, as specified in the [documentation](). There is one thing to note though: For reasons of numerical stability, the actual output tensor of the model should be of type `float32`.
+
+```{r}
+model <- keras_model_sequential() %>% 
+  layer_conv_2d(filters = 32, kernel_size = 5, strides = 2, padding = "same", input_shape = c(256, 256, 3), activation = "relu") %>%
+  layer_batch_normalization() %>%
+  layer_conv_2d(filters = 64, kernel_size = 7, strides = 2, padding = "same", activation = "relu") %>%
+  layer_batch_normalization() %>%
+  layer_conv_2d(filters = 128, kernel_size = 11, strides = 2, padding = "same", activation = "relu") %>%
+  layer_batch_normalization() %>%
+  layer_global_average_pooling_2d() %>%
+  # separate logits from activations so actual outputs can be float32
+  layer_dense(units = 10) %>%
```

Comment:
Can we try a multiple of 8 for number of units?

Reply:
there are 10 classes, so, no... also, I think given this is just a single layer after several conv layers, so even if this made a difference, it shouldn't be a big one...

btw I also tried without batchnorm, this didn't make a difference either

Reply:
LOL yeah this is the output layer, brainfart on my part